Design doc: collect data about builds (#8124)
readthedocs/readthedocs.org: docs/development/design/telemetry.rst

Collect Data About Builds
=========================

We may want to take some decisions in the future about deprecations and supported versions.

> Right now we don't have data about the usage of packages and their versions on Read the Docs

Reviewer: Moarrrrrrrrrr data! 🤓

Reviewer: Now, more seriously: aren't we storing the build logs already somewhere? I think this proposal is more about storing **structured** data.

Author: build logs don't include the output from `pip list` or `conda list`

Author: but yeah, we already save the build output in json files in s3

> to be able to make a good decision.

Reviewer: "Good" or "bad" is a bit ambiguous I guess, what about

```suggestion
to be able to make an informed decision.
```

.. contents::
   :local:
   :depth: 3

Data to be collected
--------------------

The following data can be collected after installing all dependencies.

Configuration file
~~~~~~~~~~~~~~~~~~

We are saving the config file in our database,
but to save some space we are saving it only if it's different than the one from a previous build
(if it's the same we save a reference to it).

The config file being saved isn't the original one used by the user,
but the result of merging it with its default values.
It's saved using a _fake_ ``JSONField``
(charfield that is transformed to json when creating the model object).

> For these reasons we can't query or download them in bulk without iterating over all objects.

Reviewer: I see how this is a problem, but it's not clear to me what the solution would be?

Reviewer: Related: perhaps only certain keys of the configuration are of interest and this simplifies the problem? But I have no idea how are we parsing the `conf.py` at the moment.

Reviewer 2: I think we need to upgrade Django to get real JSON fields, which let us query them.

Author: >  but it's not clear to me what the solution would be?

Use real json fields or another form to store this information

We may also want to have the original config file,
so we know which settings users are using.

PIP packages
~~~~~~~~~~~~

We can get a json with all root dependencies with ``pip list``.
This will allow us to have the name of the packages and their versions used in the build.

.. code-block::

>    $ pip list --pre --not-required --local --format json | jq

Reviewer: About `--not-required`: even if it's more space, aren't we interested in the transitive dependencies as well?

Reviewer: Also, TIL about `--format json` 🚀

Reviewer: The good thing about listing only the not required ones is that this gives us an idea of what was specified. But perhaps we should remove `--not-required` and try to store what the user specified instead: the reason is that we also want to know who pins their dependencies, and who doesn't.

Reviewer 2: We probably want both. For example, the version of docutils will likely be decided by transitive requirements (eg. our theme) and we don't want full dependency resolution code on our side :)

I think we can store the user-defined requirements, and also the full set.

   [
      {
         "name": "requests-mock",
         "version": "1.8.0"
      },
      {
         "name": "requests-toolbelt",
         "version": "0.9.1"
      },
      {
         "name": "rstcheck",
         "version": "3.3.1"
      },
      {
         "name": "selectolax",
         "version": "0.2.10"
      },
      {
         "name": "slumber",
         "version": "0.7.1"
      },
      {
         "name": "sphinx-autobuild",
         "version": "2020.9.1"
      },
      {
         "name": "sphinx-hoverxref",
         "version": "0.5b1"
      },
   ]

Conda packages
~~~~~~~~~~~~~~

We can get a json with all dependencies with ``conda list --json``.
That command gets all the root dependencies and their dependencies,

> so we may be collecting some noise, but we can use ``pip list`` as a secondary source.

Reviewer: Similarly with pip, yes this gives us all installed dependencies without distinction of whether they are transitive dependencies or not (too bad), but probably we should also try to store the `environment.yml` somehow, just an idea.

That command would list all pre-installed packages as well, so we may be getting some noise.

.. code-block::

   $ dpkg --get-selections

   adduser                                         install
   apt                                             install
   base-files                                      install
   base-passwd                                     install
   bash                                            install
   binutils                                        install
   binutils-common:amd64                           install
   binutils-x86-64-linux-gnu                       install
   bsdutils                                        install
   build-essential                                 install

Python
~~~~~~

We can get the Python version from the config file when using a Python environment,
and from the ``conda list`` output when using a Conda environment.

OS
~~

We can infer the OS version from the build image used in the config file,
but since it changes with time, we can get it from the OS itself:

.. code-block::

   $ lsb_release --description
   Description:    Ubuntu 18.04.5 LTS
   # or
   $ cat /etc/issue
   Ubuntu 18.04.5 LTS \n \l

> 

Reviewer: Another thing I'm missing: running times of each build step.

Storage
-------

> We can save all this information in json files in cloud storage,

Reviewer: I'm not an expert, but I believe ElasticSearch is a great fit for this kind of data, and we are already using it for search.

Reviewer 2: The main issue is that we quickly run out of memory storing this in anything that is "live". We don't want to query this data in any kind of real-time manner, so something closer to a map reduce type workflow is probably best. JSON in cloud storage probably isn't ideal for this, but I'm pretty confident we don't want it stored "hot" unless we're only storing a short timeframe. 

A couple options here:

* Only store the last 5 builds for any version
* Only store the last 3 months of builds

There are tradeoffs here, but if we scale the amount of data down, it gives us more flexibility.

Reviewer 3: I'd also considering storing only 1 build per active version. I don't think we want historical data for this, but the actual one for the current active versions.

Reviewer 4: Another +1 for avoiding long term hot storage of metrics. S3/cold storage isn't great to work with when doing analysis of build/build commands, but we can address this with some scripts to automate the process. 1 month of build data is about 7G of cloned files, but luckily we're not going to inspect this data frequently.

then we could use a tool to import all this data into.
Or we can decide for a tool or service where to fed all this data directly into.

If we decide to save the files in cloud storage,
we can try to calculate a hash of the file so we don't upload duplicates that happen on the same day/month.
We can aggregate this data per year/month saving them in following structure:
``telemetry/builds/{year}/{month}/{year}-{month}-{day}-{timestamp-pk|pk}.json``,
that way is easy to download, all data per year/month without iterating over all files.

.. Since this information isn't sensitive,
   I think we are fine with this structure

>    (we can't do bulk deletes of all info about a project if we follow this structure).

Reviewer 2: I think storing it in "cold" storage is right for the long-term large-scale analysis. I'd love to have a subset of important things look more like our monitoring data, and reported/graphed live via cloudwatch. We could pick the top 10 most important/interesting things (eg. theme, sphinx version, etc.)

Format
~~~~~~

The final file to be saved would have the following information:

- project: the project slug
- version: the version slug
- build: the build id (which may stop existing if the project is deleted)
- date: full date in isoformat or timestamp (POSIX)
- user_config: Original user config file
- final_config: Final configuration used (merged with defaults)
- packages.pip: List of pip packages with name and version
- packages.conda: List of conda packages with name, channel, and version
- packages.apt: List of apt packages
- python: Python version used
- os: Operating system used

.. code-block:: json

   {
     "project": "docs",
     "version": "latest",
     "build": 12,
     "date": "2021-04-20-...",
     "user_config": {},
     "final_config": {},
     "packages": {
        "pip": [{
           "name": "sphinx",
           "version": "3.4.5"
        }],
        "conda": [{
           "name": "sphinx",
           "channel": "conda-forge",
           "version": "0.1"

>         }],

Author: I think we can remove the root packages from this list or merge them into one list like:

```json
{"name": "pip", "version": "10.0", "root": false}
```

        "apt": [
            "python3-dev",
            "cmatrix"
        ]
     },
     "python": "3.7",
     "os": {
         "name": "ubuntu",
         "version": "18.04.5"
     }
   }

Analyzing the data
------------------

> 

Reviewer 2: I think we actually probably need to have this section first. Knowing what kind of analysis we want to run on the data will tell us how to store & query it. 

Currently we have defined a way of storing data that we don't really look at. I think it's probably more valuable to start to have some kind of short-term storage of data in a query engine that we can query dynamically (ES or Postgres, as @astrojuanlu said). We'd have to pretty strongly truncate this data (eg. <4GB in memory) -- but I think we can definitely get some useful subsets of data into that format given that it isn't huge.

In particular, just getting this data into some kind of queryable format:

* .readthedocs.yml
* conf.py dumped as JSON
* pip & conda files dumped

Would get us a lot of value across a short timeframe (eg. 1 month).

But, what are we going to do when we have this data to query? We should think about what kind of queries we want to run and what analysis would be valuable to make sure we're building the system properly.

   Description:    Ubuntu 18.04.5 LTS
   # or
   $ cat /etc/issue
   Ubuntu 18.04.5 LTS \n \l

Format
~~~~~~

The final file to be saved would have the following information:

- project: the project slug
- version: the version slug
- build: the build id (which may stop existing if the project is deleted)
- date: full date in isoformat or timestamp (POSIX)
- user_config: Original user config file
- final_config: Final configuration used (merged with defaults)
- packages.pip: List of pip packages with name and version
- packages.conda: List of conda packages with name, channel, and version
- packages.apt: List of apt packages
- python: Python version used
- os: Operating system used

.. code-block:: json

   {
     "project": "docs",
     "version": "latest",
     "build": 12,
     "date": "2021-04-20-...",
     "user_config": {},
     "final_config": {},
     "packages": {
        "pip": [{
           "name": "sphinx",
           "version": "3.4.5"
        }],
        "pip_all": [
          {
             "name": "sphinx",
             "version": "3.4.5"
          },
          {
             "name": "docutils",
             "version": "0.16.0"
          }

>         ],

Author: I think we can remove the root packages from this list or merge them into one list like:

```json
{"name": "pip", "version": "10.0", "root": false}
```
